### PR TITLE
Island: Add concrete AgentBinaryRepository

### PR DESCRIPTION
**Pull request by ilija-lazoroski**
# What does this PR do?

Implements concrete AgentBinaryReposiotyr, #1974 .

Add any further explanations here.

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [x] Is the TravisCI build passing?
    > vulture failing until the Agent-binaries resource is refactored
* [x] Was the CHANGELOG.md updated to reflect the changes?
* [x] Was the documentation framework updated to reflect the changes?
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Have you successfully tested your changes locally? Elaborate:
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [x] If applicable, add screenshots or log transcripts of the feature working

## Explain Changes

Are the commit messages enough? If not, elaborate.

### CODE REVIEW

File: monkey/monkey_island/cc/repository/agent_binary_repository.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+            logger.error(f"An error has occured retriving linux agent binary: {err}")
+            raise AgentRetrivalError(f"It couldn't retrieve linux agent binary: {err}")
+
+    def get_windows_binary(self) -> BinaryIO:
+        try:
+            agent_binary = self._file_repository.open_file(AGENTS["windows"])
+            return agent_binary
+        except FileRetrivalError as err:
+            logger.error(f"An error has occured retriving windows agent binary: {err}")
+            raise AgentRetrivalError(f"It couldn't retrieve windows agent binary: {err}")
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/repository/agent_binary_repository.py:29]:
```suggestion
            raise AgentRetrivalError(f"An error occurred while retrieving the Windows agent binary: {err}") from err
```

File: monkey/monkey_island/cc/repository/agent_binary_repository.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    def __init__(self, file_repository: IFileRepository):
+        self._file_repository = file_repository
+
+    def get_linux_binary(self) -> BinaryIO:
+        try:
+            agent_binary = self._file_repository.open_file(AGENTS["linux"])
+            return agent_binary
+        except FileRetrivalError as err:
+            logger.error(f"An error has occured retriving linux agent binary: {err}")
+            raise AgentRetrivalError(f"It couldn't retrieve linux agent binary: {err}")
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/repository/agent_binary_repository.py:21]: Don't log and raise. Just raise. Whatever catches this exception can choose whether or not to log it.

```suggestion
            raise AgentRetrivalError(f"An error occurred while retrieving the Linux agent binary: {err}") from err
```

**Review by mssalvatore — CHANGES_REQUESTED** [review]

File: monkey/monkey_island/cc/repository/agent_binary_repository.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+    def get_windows_binary(self) -> BinaryIO:
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/repository/agent_binary_repository.py:23]: This whole method is almost the exact same as the `get_linux_binary`

File: monkey/monkey_island/cc/repository/agent_binary_repository.py (hunk 2 above)
Comment (VakarisZ) [reply monkey/monkey_island/cc/repository/agent_binary_repository.py:21]: `raise AgentRetrivalError("An error occurred while retrieving the Linux agent binary") from err`

File: monkey/monkey_island/cc/repository/agent_binary_repository.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import logging
+from typing import BinaryIO
+
+from . import AgentRetrivalError, FileRetrivalError, IAgentBinaryRepository, IFileRepository
+
+logger = logging.getLogger(__name__)
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/repository/agent_binary_repository.py:6]: No need to log anything in this file

File: monkey/monkey_island/cc/repository/agent_binary_repository.py (hunk 3 above)
Comment (ilija-lazoroski) [reply monkey/monkey_island/cc/repository/agent_binary_repository.py:23]: https://github.com/guardicore/monkey/pull/1986#discussion_r890238369

File: monkey/monkey_island/cc/repository/agent_binary_repository.py (hunk 2 above)
Comment (mssalvatore) [reply monkey/monkey_island/cc/repository/agent_binary_repository.py:21]: @ilija-lazoroski 
image

@VakarisZ I made the same mistake initially, but you don't need `from err`.

> Exception chaining happens automatically when an [except](https://docs.python.org/3/reference/compound_stmts.html#except)ion is raised inside an except or [finally](https://docs.python.org/3/reference/compound_stmts.html#finally) section.
https://docs.python.org/3/tutorial/errors.html

File: monkey/monkey_island/cc/repository/i_agent_binary_repository.py [hunk 5]
```diff
@@ -2,6 +2,10 @@
 from typing import BinaryIO
 
 
+class AgentRetrivalError(IOError):
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/repository/i_agent_binary_repository.py:5]: Use `RuntimeError`, not `IOError`. See https://github.com/guardicore/monkey/pull/1987#discussion_r891098724

**Review by mssalvatore — APPROVED** [review]

File: monkey/monkey_island/cc/repository/agent_binary_repository.py (hunk 3 above)
Comment (VakarisZ) [reply monkey/monkey_island/cc/repository/agent_binary_repository.py:22]: I don't see how is that comment related. This is a concrete implementation, you can create a private method:
```
def __get_binary(self, filename) -> BinaryIO:
        try:
            agent_binary = self._file_repository.open_file(filename)
            return agent_binary
        except FileRetrivalError as err:
            raise AgentRetrivalError(
                f"An error occurred while retrieving the {filename} agent binary: {err}"
            )
```
Boom, there's no more duplication

**Pull request merged**